src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -0,0 +1,179 @@
+     generated. Otherwise, no error should be generated.
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { ValidationTest } from '../validation_test.js';
+
+const READ_BASE_LEVEL = 3;
+const READ_BASE_LAYER = 0;
+
+class F extends ValidationTest {
+  createTexture(
+    options: {
+      width?: number;
+      height?: number;
+      arrayLayerCount?: number;
+      mipLevelCount?: number;
+      sampleCount?: number;
+      format?: GPUTextureFormat;
+    } = {}
+  ): GPUTexture {
+    const {
+      width = 32,
+      height = 32,
+      arrayLayerCount = 1,
+      mipLevelCount = 1,
+      sampleCount = 1,
+      format = 'rgba8unorm',
+    } = options;
+
+    return this.device.createTexture({
+      size: { width, height, depth: arrayLayerCount },
+      mipLevelCount,
+      sampleCount,
+      dimension: '2d',
+      format,
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT | GPUTextureUsage.SAMPLED,
+    });
+  }
+
+  getDescriptor(
+    options: {
+      format?: GPUTextureFormat;
+      dimension?: GPUTextureViewDimension;
+      aspect?: GPUTextureAspect;
+      baseMipLevel?: number;
+      mipLevelCount?: number;
+      baseArrayLayer?: number;
+      arrayLayerCount?: number;
+    } = {}
+  ): GPUTextureViewDescriptor {
+    const {
+      format = 'rgba8unorm',
+      dimension = '2d',
+      aspect = 'all',
+      baseMipLevel = 0,
+      mipLevelCount = 1,
+      baseArrayLayer = 0,
+      arrayLayerCount = 1,
+    } = options;
+    return {
+      format,
+      dimension,
+      aspect,
+      baseMipLevel,
+      mipLevelCount,
+      baseArrayLayer,
+      arrayLayerCount,
+    };
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('readwrite_on_different_types_of_subresources')
+  .params([
+    // read and write usages are binding to the same texture subresource.
+    {
+      format: 'rgba8unorm' as const,
+      writeBaseLevel: READ_BASE_LEVEL,
+      writeBaseLayer: READ_BASE_LAYER,
+      writeAspect: 'all' as const,
+      success: false,
+    },
+
+    // read and write usages are binding to different mip levels of the same texture.
+    {
+      format: 'rgba8unorm' as const,
+      writeBaseLevel: READ_BASE_LEVEL + 1,
+      writeBaseLayer: READ_BASE_LAYER,
+      writeAspect: 'all' as const,
+      success: true,
+    },
+
+    // read and write usages are binding to different array layers of the same texture.
+    {
+      format: 'rgba8unorm' as const,
+      writeBaseLevel: READ_BASE_LEVEL,
+      writeBaseLayer: READ_BASE_LAYER + 1,
+      writeAspect: 'all' as const,
+      success: true,
+    },
+
+    // read and write usages are binding to different aspects of the same texture.
+    {
+      format: 'depth24plus-stencil8' as const,
+      writeBaseLevel: READ_BASE_LEVEL,
+      writeBaseLayer: READ_BASE_LAYER,
+      writeAspect: 'stencil-only' as const,
+      success: true,
+    },
```

Comment:
Need to more thoroughly test the depth/stencil cases:
- DS format, both 'all'
- DS format, both 'depth'
- DS format, both 'stencil'
- DS format, read 'stencil', write 'depth'
- DS format, read 'depth', write 'stencil'
- DS format, read 'all', write 'depth'
- DS format, read 'all', write 'stencil'
- DS format, read 'stencil', write 'all'
- DS format, read 'depth', write 'all'
- S format, both 'all'
- S format, both 'stencil'
- S format, read 'stencil', write 'all'
- S format, read 'all', write 'stencil'
- D format, both 'all'
- D format, both 'stencil'
- D format, read 'stencil', write 'all'
- D format, read 'all', write 'stencil'

i.e. something like:

```ts
params()
  .combine(poptions('format', [ds, s, d]))
  .combine(poptions('readAspect', ['all', 'depth-only', 'stencil-only']))
  .combine(poptions('writeAspect', ['all', 'depth-only', 'stencil-only']))
  .unless(format doesn't have depth and either aspect is 'depth-only')
  .unless(format doesn't have stencil and either aspect is 'stencil-only')
```

Reply:
I assume "S format" you mentioned means stencil-only format. It looks like there is no stencil-only format. The format we can support for depth/stencil are     
    "depth32float",
    "depth24plus",
    "depth24plus-stencil8", 
I suppose the first two are depth-only. The last one is depth-stencil. So there is no stencil-only. WDYT? @kainino0x

Reply:
right, sorry, I forgot - we decided to add one in gpuweb/gpuweb#306 but it's not done yet. Since we will add it, let's make sure to set up the test so it can handle adding an additional format to the list. Either add TODO to test it once it's in, or parameterize it with a const array defined in capability_info that lists all the D/S/DS formats.

OK to make it a different test as you see fit.

Reply:
Done. Added a separate test for depth/stencil, add added a TODO for stencil-only format